Add SPI to decode image data and get information

szewai opened the pull request:
#### 28f45e31db3ad8f23bbc11dcff239aa2c1d7eb00
<pre>
Add SPI to decode image data and get information
<a href="https://bugs.webkit.org/show_bug.cgi?id=281257">https://bugs.webkit.org/show_bug.cgi?id=281257</a>
<a href="https://rdar.apple.com/136705471">rdar://136705471</a>

Reviewed by Said Abou-Hallawa.

Provide a way for client to decode image data in a sandboxed process.
New API Test: WebKit.GetInformationFromImageData.

* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::descriptionString):
(WebCore::utiAndAvailableSizesFromImageData):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::isDataDecodable):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getInformationFromImageData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getInformationFromImageData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getInformationFromImageData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, GetInformationFromImageData)):

Canonical link: <a href="https://commits.webkit.org/285439@main">https://commits.webkit.org/285439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f4169a957127ebfcdd8fe5a99500dda86c74982

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25420 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76811 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23656 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47091 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62527 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/72129 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43741 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19994 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22185 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65602 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78481 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19487 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16916 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62536 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64863 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16044 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13161 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47845 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48912 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50207 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48657 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 55;
```
Most of these changes should not be included in this PR.

szewai replied:
Will revert.

achristensen07 commented on Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
```diff
@@ -46,6 +46,7 @@
 #import <wtf/RuntimeApplicationChecks.h>
 #import <wtf/URL.h>
 #import <wtf/cocoa/NSURLExtras.h>
+#import <wtf/cocoa/RuntimeApplicationChecksCocoa.h>
```
This change seems unrelated.  Is it necessary to fix unified builds after this?

szewai replied:
Yes.

achristensen07 commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+#endif
+
+#if !PLATFORM(MAC)
+SOFT_LINK_PRIVATE_FRAMEWORK(CoreSVG)
```
Can this be put into a soft link header and cpp?

achristensen07 commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+
+        float width = 0, height = 0;
+        CFNumberRef cfWidth = (CFNumberRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyPixelWidth);
+        CFNumberGetValue(cfWidth, kCFNumberFloatType, &width);
```
null check the CFNumber?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -156,4 +156,33 @@
+
+    done = false;
+    RetainPtr svgData = [NSData dataWithContentsOfURL:[NSBundle.test_resourcesBundle URLForResource:@"AllAhem" withExtension:@"svg"]];
+    [webView _getInformationFromImageData:svgData.get() completionHandler:^(BOOL canBeDecodedAsSVG, NSArray<NSValue *> *availableSizes, NSError *error) {
```
It seems like this is the first of two steps, the second of which would be actually getting the bitmap or svg.  Is that correct?

szewai replied:
Yes, we will have a different SPI for getting the actual image. This SPI is only about extracting information so that client can decide whether to proceed.

darinadler commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    if (!parser || ![parser parse])
+        return false;
+
+    CGSVGDocumentRef document = CGSVGDocumentCreateFromData((__bridge CFDataRef)nsData.get(), nullptr);
```
Do we need to use `CGSVGDocumentRelease` or can we just use `adoptCF` for this instead?

We have `bridge_cast`, which is better than writing `(__bridge CFDataRef)`.

szewai replied:
I updated the function to use `DOMParser` instead of CoreSVG as we may want to use WebKit's own parser for rendering image later, so this part is removed.

darinadler commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+        return Vector<FloatSize> { };
+
+    RetainPtr nsData = toNSData(data);
+    auto source = adoptCF(CGImageSourceCreateWithData((__bridge CFDataRef)nsData.get(), nullptr));
```
We have `bridge_cast`, which is better than writing `(__bridge CFDataRef)`.

szewai replied:
I updated the function to use `ImageDecoderCG` per Said's advice, so this part is removed.

darinadler commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+        }
+
+        float width = 0, height = 0;
+        CFNumberRef cfWidth = (CFNumberRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyPixelWidth);
```
We have `checked_cf_cast`, which should be used in cases like this.

szewai replied:
This part is removed.

darinadler commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -44,5 +44,10 @@ WEBCORE_EXPORT Vector<String> findImagesForTranscoding(const Vector<String>& pat
 
+#if PLATFORM(COCOA)
+WEBCORE_EXPORT bool canDataBeDecodedAsSVG(std::span<const uint8_t>);
+WEBCORE_EXPORT std::optional<Vector<FloatSize>> availableSizesFromData(std::span<const uint8_t>, bool isSVGData);
```
Given this just returns an empty vector when the boolean `isSVGData` is `true`, why do we bother doing it like this? Doesn’t seem all that helpful to have a function to call just to return an empty vector. How about having the caller only call `availableSizesFromData` when `canDataBeDecodedAsSVG` is `false`?

szewai replied:
Updated patch to drop `isSVGData` parameter.

darinadler commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+        float width = 0, height = 0;
+        CFNumberRef cfWidth = (CFNumberRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyPixelWidth);
+        CFNumberGetValue(cfWidth, kCFNumberFloatType, &width);
+        CFNumberRef cfHeight = (CFNumberRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyPixelHeight);
```
We have `checked_cf_cast`, which should be used in cases like this.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3256,6 +3256,25 @@ - (void)_loadAndDecodeImage:(NSURLRequest *)request constrainedToSize:(CGSize)ma
+            return completionHandler(canBeRepresentedAsSVG, nil, error);
+        }
+
+        RetainPtr availableSizes = createNSArray(*sizes, [](auto& size) {
```
Why `auto&` instead of `auto`?

szewai replied:
Updated to use auto.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3256,6 +3256,25 @@ - (void)_loadAndDecodeImage:(NSURLRequest *)request constrainedToSize:(CGSize)ma
+#if PLATFORM(MAC)
+            return [NSValue valueWithSize:size];
+#else
+            return [NSValue valueWithCGSize:(CGSize)size];
```
Is there a cleaner way to convert to `CGSize` instead of a C style cast? Maybe just a local variable?

szewai replied:
Updated to use local variable.

shallawa commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -44,5 +44,10 @@ WEBCORE_EXPORT Vector<String> findImagesForTranscoding(const Vector<String>& pat
 
+#if PLATFORM(COCOA)
+WEBCORE_EXPORT bool canDataBeDecodedAsSVG(std::span<const uint8_t>);
+WEBCORE_EXPORT std::optional<Vector<FloatSize>> availableSizesFromData(std::span<const uint8_t>, bool isSVGData);
```
I do not understand this interface fully. I think you want to get

1. Whether the encoded data is a bitmap image or SVG image
2. Some metadata about the bitmap image. Here you want to get the size(s) of image frame(s).

I think the first API should say this explicitly like `isSVGImage` or `isSVGImageRep`.
The second API is a little problematic. I think it should be used solely for bitmap image. It does not make sense to send to it the flag `isSVGData` and if it is true it returns an empty `Vector<FloatSize>`. We can do this without calling the function:
```
auto sizes = isSVGData ? Vector<FloatSize>() : availableSizesForBitmapImage(data);
```

szewai replied:
Updated patch as suggested.

shallawa commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    if (!source)
+        return std::nullopt;
+
+    size_t count = CGImageSourceGetCount(source.get());
```
There is a problem if the image is an animated image. If the loop count is zero. This code is correct. Otherwise  I do not know what available size for an animated image are.

I know some image formats like HEIC can hold multiple frames with different size. But this should not be confused with animated images where all the frames are different.

szewai replied:
This is a good point. I've updated the function to exclude the case for animated image. Could you take a look again?

shallawa commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    RetainPtr nsData = toNSData(data);
+    auto source = adoptCF(CGImageSourceCreateWithData((__bridge CFDataRef)nsData.get(), nullptr));
+    if (!source)
+        return std::nullopt;
```
Do not we need to know what went wrong when creating the `CGImageSource`? 
Do not we need to know whether the encoded data is of the supported image formats: `isSupportedImageType()`? We are not supposed to decode an image which is neither in `defaultSupportedImageTypes()` nor `additionalSupportedImageTypes()` because of security reasons.

szewai replied:
Thanks for pointing this out. I added type check in the function; also I made the SPI return UTI so client can tell if data type is not image.

shallawa commented on Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    Vector<FloatSize> sizes;
+    sizes.reserveInitialCapacity(count);
+    for (size_t index = 0; index < count; ++index) {
+        auto properties = adoptCF(CGImageSourceCopyPropertiesAtIndex(source.get(), index, nullptr));
```
Can't we reuse some of the WebCore code instead of repeating it here: `ImageDecoderCG` or `BitmapImage`?

szewai replied:
Yes, I updated the patch to use `ImageDecoderCG` for this function.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -645,7 +645,8 @@ class Document
     bool isPDFDocument() const { return m_documentClasses.contains(DocumentClass::PDF); }
-    bool hasSVGRootNode() const;
+    WEBCORE_EXPORT bool hasSVGRootNode() const;
+    WEBCORE_EXPORT bool mayBeSVGData(const Vector<uint8_t>&);
```
This should take `std::span<const uchar8_t>`, `std::span<const uint8_t>`, `const String&`, or `StringView` instead of `const Vector&`. There’s no reason to require that the caller store the data in a vector specifically.

Also, given that this doesn’t use any `Document` state, this should be a static member function if it’s in the `Document` class. But very unclear why it’s in the `Document` class. Doesn’t belong here. Let’s find a better place for it.

szewai replied:
`DOMParser` needs to be created with `Document` and Document.cpp has necessary headers for the functions so I put it there. Do you think it's better to create a helper function like below and move it to other places?

darinadler replied:
A function to check if some data can be parsed as SVG data should be in a header about that’s about SVG, parsing, or maybe file types. The implementation detail that we parse into a document and then check the type of node should **not** determine where the function goes. I don’t immediately know which header file, but it shouldn’t be too hard to find a better one than `Document`, which is included by literally everything and should not be unnecessarily expanded to include new things.

szewai replied:
Got it; moved the function to Source/WebCore/platform/graphics/ImageUtilities.h instead.

darinadler commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -29,6 +29,7 @@
 #include <wtf/WorkQueue.h>
 
 namespace WebCore {
+class IntSize;
```
Need a blank line before this.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +125,24 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+
+    // No support for animated image.
+    if (imageDecoder->repetitionCount() == RepetitionCountInfinite)
+        return { uti, { } };
```
Comment doesn’t match the code; images with repetition counts like 2 are also animated images.

szewai replied:
I think repetition count 2 is infinite since we only have None, Once and Infinite:
https://github.com/WebKit/WebKit/blob/f607bbdba4df616b144d16367b4f4fa3ee329a02/Source/WebCore/platform/graphics/ImageTypes.h#L55

darinadler replied:
I believe `Once` is also an animated image.

szewai replied:
I found gif could have 1 frame with loop count 1 and I think we need to compute size for that case. To make it clear, I updated this check to return early if loop count is not 0 and it contains multiple frames.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +125,24 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    Ref buffer = FragmentedSharedBuffer::create(data);
+    Ref imageDecoder = ImageDecoderCG::create(buffer.get(), AlphaOption::Premultiplied, GammaAndColorProfileOption::Applied);
+    imageDecoder->setData(buffer.get(), true);
+    auto uti = imageDecoder->uti();
```
This local variable is never used.

szewai replied:
ah nice catch! Will fix.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +125,24 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    if (imageDecoder->repetitionCount() == RepetitionCountInfinite)
+        return { uti, { } };
+
+    Vector<IntSize> sizes;
```
We know the capacity of this vector will be the frame count, so we should reserve the initial capacity for better efficiency.

darinadler commented on Source/WebCore/xml/DOMParser.h:
```diff
@@ -30,10 +30,10 @@ class TrustedHTML;
+    WEBCORE_EXPORT ~DOMParser();
 
-    ExceptionOr<Ref<Document>> parseFromString(std::variant<RefPtr<TrustedHTML>, String>&&, const AtomString& contentType);
+    WEBCORE_EXPORT ExceptionOr<Ref<Document>> parseFromString(std::variant<RefPtr<TrustedHTML>, String>&&, const AtomString& contentType);
```
Why is this being exported? The new caller seems to be in `Document`, which is in WebCore.

szewai replied:
ah these changes were used for testing; will revert.

darinadler commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7975,6 +7976,20 @@ void WebPage::loadAndDecodeImage(WebCore::ResourceRequest&& request, std::option
+        return completionHandler("public.svg-image"_s, { });
+
+    auto [uti, sizes] = utiAndAvailableSizesFromImageData(data.span());
+    completionHandler(uti, WTFMove(sizes));
```
Should have `WTFMove` on `uti` too.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -187,6 +187,7 @@ typedef NS_ENUM(NSInteger, _WKImmediateActionType) {
 #else
 - (void)_loadAndDecodeImage:(NSURLRequest *)request constrainedToSize:(CGSize)maxSize maximumBytesFromNetwork:(size_t)maximumBytesFromNetwork completionHandler:(void (^)(NSImage *, NSError *))completionHandler WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
 #endif
+- (void)_getInformationFromImageData:(NSData *)imageData completionHandler:(void (^)(NSString *, NSArray<NSValue *> *availableSizes, NSError *error))completionHandler;
```
It’s not obvious what the `NSString` is without an argument name. We also need to add `WK_API_AVAILABLE` for this.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -645,7 +645,8 @@ class Document
 #endif
     bool isPDFDocument() const { return m_documentClasses.contains(DocumentClass::PDF); }
-    bool hasSVGRootNode() const;
+    WEBCORE_EXPORT bool hasSVGRootNode() const;
```
Don’t think this needs `WEBCORE_EXPORT`.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7975,6 +7976,20 @@ void WebPage::loadAndDecodeImage(WebCore::ResourceRequest&& request, std::option
+void WebPage::getInformationFromImageData(const Vector<uint8_t>& data, CompletionHandler<void(String, Vector<WebCore::IntSize>&&)>&& completionHandler)
+{
+#if PLATFORM(COCOA)
+    auto document = Document::create(m_page->settings(), aboutBlankURL());
```
It's a bit odd to create an `about:blank` document then parse SVG in the context of that document; I'm surprised we don't run into XML namespace issues. Can we instead have the document be the result of parsing the buffer?

szewai replied:
Is there a function to create `Document` from a string? I only found `DOMParser` and it requires a container `Document` to work.

smfr commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +125,26 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+    Vector<IntSize> sizes;
+    sizes.reserveInitialCapacity(frameCount);
+    for (size_t index = 0; index < frameCount; ++index)
+        sizes.append(imageDecoder->frameSizeAtIndex(index));
```
If this is a huge animated GIF, getting the size of every frame will be very expensive, and I bet the client only cares about multiple frame sizes for multi-image HEICs. Perhaps we can limit this to only multi-representation images?

szewai replied:
Yes I don't think we need to get size of all frames for animated GIF, so I added check for loop count in the function.

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8828,6 +8828,15 @@ void WebPageProxy::loadAndDecodeImage(WebCore::ResourceRequest&& request, std::o
+{
+    if (!hasRunningProcess())
+        launchProcess(Site(aboutBlankURL()), ProcessLaunchReason::InitialProcess);
+    sendWithAsyncReply(Messages::WebPage::GetInformationFromImageData(WTFMove(data)), [preventProcessShutdownScope = protectedLegacyMainFrameProcess()->shutdownPreventingScope(), completionHandler = WTFMove(completionHandler)] (String typeIdentifier, auto availableSizes) mutable {
```
Is the API client expected to be careful to use a separate WKWebView for these queries? If they use an existing WKWebView, the image load will happen in the existing process, which could lead to some security challenges.

szewai replied:
Yes, for better security they should use a separate view for loading these image decoding requests, though the SPI does not have enforcement that this has to be a blank view (as the client may use it for different requests).

shallawa commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -30,6 +30,9 @@
 namespace WebCore {
 
+class IntSize;
+class Settings;
```
I think this is layering variation. Settings should not be referenced in the platform directory.

szewai replied:
Moved to `SVGImage`.

shallawa commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -44,5 +47,8 @@ WEBCORE_EXPORT Vector<String> findImagesForTranscoding(const Vector<String>& pat
 WEBCORE_EXPORT Vector<String> transcodeImages(const Vector<String>& paths, const String& destinationUTI, const String& destinationExtension);
 
+WEBCORE_EXPORT std::pair<String, Vector<IntSize>> utiAndAvailableSizesFromImageData(std::span<const uint8_t>);
+WEBCORE_EXPORT bool mayBeSVGData(const Settings&, std::span<const uint8_t>);
```
I think maybe is an adverb that means "possibly" or "perhaps". So I think it should treated as one word.

szewai replied:
Updated.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +128,37 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+{
+    Ref buffer = FragmentedSharedBuffer::create(data);
+    Ref imageDecoder = ImageDecoderCG::create(buffer.get(), AlphaOption::Premultiplied, GammaAndColorProfileOption::Applied);
+    imageDecoder->setData(buffer.get(), true);
```
In case you want to differentiate between the case of decoding error and unsupported image types, you can check for the encodedDataStatus() here:
```
if (imageDecoder->encodedDataStatus() == EncodedDataStatus::Error)
    // return a decoding error here
```

szewai replied:
Added error.

shallawa commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -44,5 +47,8 @@ WEBCORE_EXPORT Vector<String> findImagesForTranscoding(const Vector<String>& pat
 // happens while transcoding, a null string will be added to the returned list.
 WEBCORE_EXPORT Vector<String> transcodeImages(const Vector<String>& paths, const String& destinationUTI, const String& destinationExtension);
 
+WEBCORE_EXPORT std::pair<String, Vector<IntSize>> utiAndAvailableSizesFromImageData(std::span<const uint8_t>);
```
Should not it return an `std::optional<std::pair<String, Vector<IntSize>>>` so it can return `std::nullopt` if an error happens?

We can also return 
```
enum class DecodingError {
    None,
    DataError,
    UnsupportedType
};

Expected<std::pair<String, Vector<IntSize>, DecodingError>>
```

szewai replied:
Updated return value type to `Expected<std::pair<String, Vector<IntSize>, ImageDecodingError>>`

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -125,4 +128,37 @@ Vector<String> transcodeImages(const Vector<String>& paths, const String& destin
+bool mayBeSVGData(const Settings& settings, std::span<const uint8_t> data)
+{
+    auto document = Document::create(settings, aboutBlankURL());
+    auto domParser = DOMParser::create(document.get());
```
This is another layering violation. `Document` and `DOMParser` should not be used in the platform directory. I would suggest moving this function to SVGImage.cpp. It can be even a static function in `SVGImage` class. It can be named `SVGImage::isDataDecodable()` or `SVGImage::canDataBeDecoded()` or something. But this is not necessary if you want to keep it a standalone function.

shallawa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -156,4 +159,46 @@
+        for (NSValue *size in availableSizes) {
+            EXPECT_EQ(215, [size sizeValue].width);
+            EXPECT_EQ(174, [size sizeValue].height);
+        }
```
Why do we have to loop when the count is 1?

szewai replied:
To avoid crash if `availableSizes.count` is not 1; I updated to use `firstObject`

shallawa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -156,4 +159,46 @@
+        done = true;
+    }];
+    Util::run(&done);
+}
```
Should we add a test case for unsupported image types like JPEG2000 for example?

szewai replied:
Yes, added

szewai committed:
Add SPI to decode image data and get information
https://bugs.webkit.org/show_bug.cgi?id=281257
rdar://136705471

Reviewed by Said Abou-Hallawa.

Provide a way for client to decode image data in a sandboxed process.
New API Test: WebKit.GetInformationFromImageData.

* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::descriptionString):
(WebCore::utiAndAvailableSizesFromImageData):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::isDataDecodable):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getInformationFromImageData:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getInformationFromImageData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getInformationFromImageData):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, GetInformationFromImageData)):

Canonical link: https://commits.webkit.org/285439@main
